src/ucp/tag/rndv.c

```diff
@@ -637,11 +725,17 @@ static void ucp_rndv_prepare_zcopy_send_buffer(ucp_request_t *sreq, ucp_ep_h ep)
         (ucp_ep_get_am_lane(ep) != ucp_ep_get_tag_lane(ep))) {
         ucp_request_send_buffer_dereg(sreq, ucp_ep_get_tag_lane(sreq->send.ep));
         sreq->send.state.dt.contig.memh = UCT_MEM_HANDLE_NULL;
-    } else if ((ucp_ep_is_rndv_lane_present(ep)) &&
-        (ucp_ep_get_am_lane(ep) != ucp_ep_get_rndv_get_lane(ep))) {
+        ucp_request_clear_rails(&sreq->send.state);
+    } else if ((ucp_ep_is_rndv_lane_present(ep, 0)) &&
+        (ucp_ep_get_am_lane(ep) != ucp_ep_get_rndv_get_lane(ep, 0))) {
         /* dereg the original send request since we are going to send on the AM lane next */
         ucp_rndv_rma_request_send_buffer_dereg(sreq);
         sreq->send.state.dt.contig.memh = UCT_MEM_HANDLE_NULL;
+        ucp_request_clear_rails(&sreq->send.state);
+    } else if (ucp_request_have_rails(&sreq->send.state)) {
+        ucp_rndv_rma_request_send_buffer_dereg(sreq);
```

Comment:
do not get this. What if am_lane is the same as one of the rails lane? Then you would not need to deregister send buffer, right?

Reply:
for multirail we have to de-register buffers to avoid resource leaks - the problem is in de-registration: we can't de-register lanes 1+, and not de-register lane 0

Reply:
why can't we do so?

Reply:
because it will require deeper refactoring of request/memory registration, which affect non-rndv functionality too